### PR TITLE
Bugfix/objects 1219 Core Service Responses

### PR DESCRIPTION
**Pull request by andreassiegel**
### What changes were proposed in this pull request?

This improves how error responses from core services are forwarded in Things Edge.

Previously, we used an error handler for the REST template that simply accepted every response and never threw an exception. This also meant that it attempted to press the received response body into the expected model. It's no surprise that failed, because an error response, of course, looks different from a success response containing all the desired fields.

The fix uses another error handler again. This handler only accepts "acceptable" responses we can handle (404 Not Found or 409 Conflict), everything else will be considered an error, and throws a `HttpStatusCodeException`.
In order to handle this exception, a new service-specific `@ControllerAdvice` exception handler is put in place.

Further improvements are:
- ConnectExceptions occurring when a service isn't available at all are handled now
- the Thing creation response in case the Thing already existed, but Metadata was upserted using `force` is changed from _409 Conflict_ to _200 OK_

### How is this patch documented?

- Javadoc
- Code comments
- Code

### How was this patch tested?

- existing unit tests pass (again)
- manual Postman tests were successful, too

#### Depends On

Nothing.

### CODE REVIEW

File: API.adoc [hunk 1]
```diff
@@ -19,7 +19,7 @@ Authorities relevant to the Things Edge API are listed below.
 
 *Available methods*
 
-* <<create1, Create - `POST /{type}`>>
+* <<create1, Create - `POST /{type}?force={force}`>>
```
Comment (wfschneider-rfid) [inline API.adoc]: I think this is more clear `?force={true|false}`

File: API.adoc [hunk 2]
```diff
@@ -75,7 +76,13 @@ scheme must of course be compatible with the underlying data store.
 :grid: all
 
 [[create1]]
-==== Create - `POST /{type}`
+==== Create - `POST /{type}?force={force}`
```
Comment (wfschneider-rfid) [inline API.adoc]: I think this is more clear `?force={true|false}`

File: src/main/java/net/smartcosmos/edge/things/service/metadata/CreateMetadataRestService.java [hunk 3]
```diff
@@ -23,6 +24,6 @@
      * @param user the user making the request
      * @return the ResponseEntity
      */
-    @Retryable
+    @Retryable(exclude = { HttpClientErrorException.class }, include = { Throwable.class })
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/edge/things/service/metadata/CreateMetadataRestService.java]: 👍  nice! getting fancy!!  😉

File: src/main/java/net/smartcosmos/edge/things/service/things/DeleteThingRestService.java [hunk 4]
```diff
@@ -19,6 +20,6 @@
      * @param user the user making the request
      * @return the ResponseEntity
      */
-    @Retryable
+    @Retryable(exclude = { HttpClientErrorException.class }, include = { Throwable.class })
```
Comment (wfschneider-rfid) [inline src/main/java/net/smartcosmos/edge/things/service/things/DeleteThingRestService.java]: This almost starts to beg for a custom annotation

**Review by ghost — APPROVED** [review]

File: src/main/java/net/smartcosmos/edge/things/service/things/DeleteThingRestService.java (hunk 4 above)
Comment (andreassiegel) [inline src/main/java/net/smartcosmos/edge/things/service/things/DeleteThingRestService.java]: We'll see where it ends. A custom annotation may be an option going forward.

So far, it's just the first step of custom configuration. Interestingly, if `exclude` is set, `include` is required.

**Pull request merged**